File: streamflowSignatures/rpkg/tests/testthat/test-qa_flags.R
# flagged_for_high_na denominator — regression for the 2026-09-04 finding.
# Mirrors julia/test/test_qa_high_na.jl and python/tests/test_qa_flags.py.

test_that("manifest is loaded from config", {
  expect_length(streamflowsignatures:::pkg_env$qa_high_na_suffixes, 16)
  expect_length(streamflowsignatures:::pkg_env$qa_high_na_scalars, 21)
})

test_that("denominator selects signature columns by name only", {
  hdr <- c("gage_id", "latitude", "NDAMS_2009", "area_normalized", "gage_type",
           "Qann_mean", "Qann_median", "Q5_pettitt_cp_year", "elasticity_static",
           "swe_max_mean", "flagged_for_qann_range", "num_water_years")
  expect_equal(high_na_denominator_columns(hdr),
               c("Qann_mean", "Qann_median", "Q5_pettitt_cp_year", "elasticity_static", "swe_max_mean"))
  bases <- paste0("base", 1:100)
  full <- c(paste0(rep(bases, each = 16), streamflowsignatures:::pkg_env$qa_high_na_suffixes),
            streamflowsignatures:::pkg_env$qa_high_na_scalars, hdr)
  expect_length(high_na_denominator_columns(full), 1621 + 5)
})

.frame <- function() {
  data.frame(
    gage_id = c("g1", "g2", "g3", "g4"),
    Qann_mean = c(1, 1, 1, NA),
    Qann_median = c(1, NaN, 1, NA),                 # NaN counts as NA
    Q5_pettitt_cp_year = c(2000, NA, NA, NA),
    elasticity_static = c(1.2, NA, NA, NA),         # registered scalar
    swe_max_mean = c(NA, NA, 3, NA),                # NA-filled family counts
    latitude = c(40, 41, 42, 43),                   # metadata: excluded
    NDAMS_2009 = c(NA, NA, NA, 3),
    MAJ_DDENS_2009 = c(NA, NA, NA, 1),
    num_water_years = c(20L, 21L, 22L, 23L),
    area_normalized = c(TRUE, TRUE, TRUE, FALSE),
    gage_type = c("Canada", "Canada", "USGS", "USGS"),
    stringsAsFactors = FALSE
  )
}

test_that("table path: NA fraction over present signature columns (production semantics)", {
  df <- .frame()
  out <- compute_qa_flags(df)
  expect_true(is.data.frame(out))
  expect_equal(nrow(out), 4)
  expect_setequal(names(out), c("flagged_for_qann_range", "flagged_for_bfi_eckhardt_range",
    "flagged_for_bfi_lynehollick_range", "flagged_for_flashiness_range", "flagged_for_tqmean_range",
    "flagged_for_d50_range", "flagged_for_elasticity_range", "flagged_for_runoff_ratio_range",
    "flagged_for_seasonal_sum", "flagged_for_percentile_order", "flagged_for_timing_order",
    "flagged_for_high_na"))
  # 1/5, 4/5, 3/5, 5/5
  expect_equal(out$flagged_for_high_na, c(FALSE, TRUE, TRUE, TRUE))
  # data.table input behaves identically
  expect_equal(compute_qa_flags(data.table::as.data.table(df))$flagged_for_high_na,
               c(FALSE, TRUE, TRUE, TRUE))
  # the 11 range flags agree with the per-gage list path row by row
  for (i in seq_len(nrow(df))) {
    row <- compute_qa_flags(as.list(df[i, , drop = FALSE]))
    for (nm in setdiff(names(out), "flagged_for_high_na")) {
      expect_identical(unname(out[[nm]][i]), unname(row[[nm]]), info = paste(nm, i))
    }
  }
})

test_that("list path: per-gage semantics over EMITTED keys", {
  sigs <- list(gage_id = "g", Qann_mean = 1, Qann_median = NaN, Q5_pettitt_cp_year = NA_real_,
               latitude = 40, NDAMS_2009 = NA_real_)   # 2 of 3 signature keys NA
  expect_true(compute_qa_flags(sigs)$flagged_for_high_na)
  sigs2 <- list(gage_id = "g", Qann_mean = 1, Qann_median = 1, Q5_pettitt_cp_year = 2000,
                NDAMS_2009 = NA_real_, MAJ_DDENS_2009 = NA_real_)   # metadata NA must not flag
  expect_false(compute_qa_flags(sigs2)$flagged_for_high_na)
})

test_that("metadata-only input never flags", {
  df <- data.frame(gage_id = c("a", "b"), latitude = c(1, 2), NDAMS_2009 = c(NA, NA))
  expect_length(high_na_denominator_columns(names(df)), 0)
  expect_equal(compute_qa_flags(df)$flagged_for_high_na, c(FALSE, FALSE))
})

File: streamflowSignatures/python/tests/test_qa_flags.py
"""flagged_for_high_na denominator — regression for the 2026-09-04 finding.

Mirrors julia/test/test_qa_high_na.jl. The denominator is defined by NAME (the 16
statistic suffixes + the per-gage scalar registry, config qa_qc.high_na_denominator);
metadata, unregistered diagnostics and flag columns never enter it. Before the fix
Python counted every non-flag column (incl. string metadata), Julia's runner path
counted metadata only, and rpkg counted mean/median keys only.
"""
import numpy as np
import pandas as pd

from streamflow_signatures.qa_qc import (
    compute_qa_flags,
    get_flag_columns,
    high_na_denominator_columns,
)
from streamflow_signatures.config import QAQC_HIGH_NA_SUFFIXES, QAQC_HIGH_NA_SCALARS
from streamflow_signatures.io import (
    ALL_STAT_SUFFIXES,
    PER_GAGE_SCALARS,
    expected_output_columns,
)


def test_manifest_matches_schema_registries():
    assert list(QAQC_HIGH_NA_SUFFIXES) == list(ALL_STAT_SUFFIXES)
    assert list(QAQC_HIGH_NA_SCALARS) == list(PER_GAGE_SCALARS)
    assert len(QAQC_HIGH_NA_SUFFIXES) == 16 and len(QAQC_HIGH_NA_SCALARS) == 21


def test_full_product_header_selects_1621_columns():
    header = expected_output_columns()
    assert len(header) == 1653
    assert len(high_na_denominator_columns(header)) == 1621
    assert not any(c.startswith("flagged_") for c in high_na_denominator_columns(header))


def _frame():
    return pd.DataFrame({
        "gage_id": ["g1", "g2", "g3", "g4"],
        "Qann_mean": [1.0, 1.0, 1.0, np.nan],
        "Qann_median": [1.0, np.nan, 1.0, np.nan],
        "Q5_pettitt_cp_year": [2000.0, np.nan, np.nan, np.nan],
        "elasticity_static": [1.2, np.nan, np.nan, np.nan],   # registered scalar
        "swe_max_mean": [np.nan, np.nan, 3.0, np.nan],        # NA-filled family counts
        # metadata: excluded from the denominator whatever their NA state
        "latitude": [40.0, 41.0, 42.0, 43.0],
        "NDAMS_2009": [np.nan, np.nan, np.nan, 3.0],
        "MAJ_DDENS_2009": [np.nan, np.nan, np.nan, 1.0],
        "num_water_years": [20, 21, 22, 23],
        "area_normalized": [True, True, True, False],
        "gage_type": ["Canada", "Canada", "USGS", "USGS"],
        "CLASS": [None, None, "Ref", None],
    })


def test_denominator_is_signature_columns_only():
    df = _frame()
    assert set(high_na_denominator_columns(df.columns)) == {
        "Qann_mean", "Qann_median", "Q5_pettitt_cp_year", "elasticity_static", "swe_max_mean"}
    out = compute_qa_flags(df)
    # NA fractions over the 5 signature columns: g1 1/5, g2 4/5, g3 3/5, g4 5/5
    assert out["flagged_for_high_na"].tolist() == [False, True, True, True]
    # a second pass must not count the flag columns themselves
    again = compute_qa_flags(out.drop(columns=get_flag_columns()))
    assert again["flagged_for_high_na"].tolist() == out["flagged_for_high_na"].tolist()


def test_metadata_only_frame_never_flags():
    df = pd.DataFrame({"gage_id": ["a", "b"], "latitude": [1.0, 2.0],
                       "NDAMS_2009": [np.nan, np.nan], "CLASS": [None, None]})
    assert high_na_denominator_columns(df.columns) == []
    assert compute_qa_flags(df)["flagged_for_high_na"].tolist() == [False, False]
